### PR TITLE
Improvement of empty lists and 404 pages

### PR DESCRIPTION
**Pull request by sipmann**
Summary of the changes (in less than 80 chars)

 * Added a message on the index/search page when there is no package on the list.
 * Added a message on the package page where the package can't be found.

Addresses https://github.com/loic-sharma/BaGet/issues/303

### CODE REVIEW

File: src/BaGet.UI/src/SearchResults.tsx [hunk 1]
```diff
@@ -156,48 +156,9 @@ class SearchResults extends React.Component<ISearchResultsProps, ISearchResultsS
...
-                }
-              </ul>
-              <div>
-                {value.description}
-              </div>
-            </div>
-          </div>
-        ))}
+
+        {this._renderSearch()}
```
Comment (loic-sharma) [inline src/BaGet.UI/src/SearchResults.tsx:160]: I'd rather avoid splitting up the HTML here if possible. Is it possible to do something like:

```tsx
{(() => {
  if (this.state.items.length > 0) {
    // Show results
  } else {
    // Show empty message
  }
})()}
```

File: src/BaGet.UI/src/SearchResults.tsx [hunk 2]
```diff
@@ -298,6 +259,68 @@ class SearchResults extends React.Component<ISearchResultsProps, ISearchResultsS
       this.state.packageType,
       this.state.targetFramework);
   }
+
+  private _renderSearch() {
+    console.log(this.state.items);
```
Comment (loic-sharma) [inline src/BaGet.UI/src/SearchResults.tsx:264]: Please remove this line

File: src/BaGet.UI/src/SearchResults.tsx [hunk 3]
```diff
@@ -298,6 +259,68 @@ class SearchResults extends React.Component<ISearchResultsProps, ISearchResultsS
...
+    {
+      return (
+        <div>
+          <h2>Oops, nothing here...</h2>
+          <p>
+            It looks like there's no package here to see. Take a look below for useful links.
+          </p>
+          <p><Link to="/upload">Upload a package</Link></p>
+          <p><a href="https://loic-sharma.github.io/BaGet/" target="_blank">BaGet Documentation</a></p>
+          <p><a href="https://github.com/loic-sharma/BaGet/issues" target="_blank">BaGet Issues</a></p>
```
Comment (loic-sharma) [inline src/BaGet.UI/src/SearchResults.tsx:319]: Casing suggestion:

```suggestion
          <p><a href="https://loic-sharma.github.io/BaGet/" target="_blank">BaGet documentation</a></p>
          <p><a href="https://github.com/loic-sharma/BaGet/issues" target="_blank">BaGet issues</a></p>
```

File: src/BaGet.UI/src/DisplayPackage/DisplayPackage.tsx [hunk 4]
```diff
@@ -205,7 +205,12 @@ class DisplayPackage extends React.Component<IDisplayPackageProps, IDisplayPacka
       );
     } else if (!this.state.package) {
       return (
-        <div>Could not find package '{this.id}'.</div>
+        <div>
+          <h2>Oops, package not found...</h2>
+          <p>Could not find package '{this.id}'.</p>
+          <p>You can take a look at the <a href={`https://www.nuget.org/packages/${this.id}`} target="_blank">nuget.org</a> package.</p>
+          <p>If you think there's a problem? Consider taking a look at our <a href="https://loic-sharma.github.io/BaGet/" target="_blank">Documentation</a> or ask for help at our <a href="https://github.com/loic-sharma/BaGet/issues" target="_blank">GitHub Project</a></p>
```
Comment (loic-sharma) [inline src/BaGet.UI/src/DisplayPackage/DisplayPackage.tsx:212]:
```suggestion
          <p>You can try searching on <a href={`https://www.nuget.org/packages?q=${this.id}`} target="_blank">nuget.org</a>.</p>
          <p>Think there's a problem? Consider taking a look at our <a href="https://loic-sharma.github.io/BaGet/" target="_blank"dDocumentation</a> or asking for help on our <a href="https://github.com/loic-sharma/BaGet/issues" target="_blank">GitHub project</a></p>
```

**Comment by loic-sharma** [comment]
Thanks for doing this, it looks great already! I left some minor comments, I'll merge this in once you address them :)

File: src/BaGet.UI/src/SearchResults.tsx (hunk 1 above)
Comment (sipmann) [reply src/BaGet.UI/src/SearchResults.tsx:160]: Yeah for sure. You mean take the code from `_renderSearch` back to the render and do the if inside it. Right?
```suggestion
      render() {
         ...
         <div className="form-group">
            <Checkbox
              defaultChecked={this.state.includePrerelease}
              onChange={this.onChangePrerelease}
              label="Include prerelease:"
              boxSide="end"
            />
          </div>
        </form>
        
        // {this._renderSearch()}
        if (this.state.items.length > 0) {
            this.state.items.map( i => ...
         } else { ...
```

Comment (sipmann) [reply src/BaGet.UI/src/SearchResults.tsx:160]: To be easier to understand. Something like this? https://gist.github.com/sipmann/9e266ca45333238ccfd9c14f033e2fa2

Comment (loic-sharma) [reply src/BaGet.UI/src/SearchResults.tsx:160]: Yup, that looks perfect

**Comment by sipmann** [comment]
Awesome, I've just commited every suggestion :). ty for your patience

File: src/BaGet.UI/src/DisplayPackage/DisplayPackage.tsx [hunk 5]
```diff
@@ -205,7 +205,12 @@ class DisplayPackage extends React.Component<IDisplayPackageProps, IDisplayPacka
       );
     } else if (!this.state.package) {
       return (
-        <div>Could not find package '{this.id}'.</div>
+        <div>
+          <h2>Oops, package not found...</h2>
+          <p>Could not find package '{this.id}'.</p>
+          <p>You can try searching on <a href={`https://www.nuget.org/packages?q=${this.id}`} target="_blank" rel="noopener noreferrer">nuget.org</a> package.</p>
```
Comment (loic-sharma) [inline src/BaGet.UI/src/DisplayPackage/DisplayPackage.tsx:211]: Nice job on these `rel` attributes, I didn't know these existed 👍

File: src/BaGet.UI/src/SearchResults.tsx [hunk 6]
```diff
@@ -156,48 +156,68 @@ class SearchResults extends React.Component<ISearchResultsProps, ISearchResultsS
...
-                onError={this.loadDefaultIcon}
-                alt="The package icon" />
-            </div>
-            <div className="col-sm-11">
-              <div>
-                <Link to={`/packages/${value.id}`} className="package-title">{value.id}</Link>
-                <span>by: {value.authors}</span>
+        
+        {(() => {
+          if (this.state.items.length > 0) {
```
Comment (loic-sharma) [inline src/BaGet.UI/src/SearchResults.tsx:161]: Hm I noticed there's a small flaw with this logic. When we change the search input, we set `items` to an empty array, thereby causing that "Oops, nothing here..." message to flash while we load the new items. I think we should introduce a new `loading` property on the state object and change this line to:

```suggestion
          if (loading || this.state.items.length > 0) {
```

When we start loading new items using [`_loadItems`](https://github.com/loic-sharma/BaGet/blob/2ed9457a0c80e4755a224911f2509f44e8e5a8bf/src/BaGet.UI/src/SearchResults.tsx#L225), we should set `loading` to `true`. Once we complete loading items, we should set `loading` to `false`.

**Comment by loic-sharma** [comment]
Ok I left one more comment to polish up the UI as we load search elements. I can't wait to get this merged in! :)

File: src/BaGet.UI/src/SearchResults.tsx (hunk 6 above)
Comment (sipmann) [reply src/BaGet.UI/src/SearchResults.tsx:161]: Yeap, totally agreed

**Comment by sipmann** [comment]
Almost done with the last suggestion. But I've found a bug at the way the search works. Fill a bug a solve in another pull request? I solve on this one?
It should be just a one line change

**Comment by loic-sharma** [comment]
What's the bug? If it's just a one liner then we can probably add it to this change as well

**Comment by sipmann** [comment]
At line 71, we are sending the previous input value, instead of the new one. So when we type the next letter, the previous one is sent.

```javascript
this._loadItems(
      prevProps.input,
```

search_bug

**Comment by loic-sharma** [comment]
Oh nice catch! Feel free to fix it in this PR or in another one, both approaches are fine by me :)

**Comment by loic-sharma** [comment]
Thanks for the awesome contribution :)

**Pull request merged**